gui/job-details.lua

```diff
@@ -0,0 +1,438 @@
+end
+
+function JobDetails:setMaterial(obj, mat_type, mat_index)
+    if  obj.index == 0
+    and self.job.mat_type == obj.iobj.mat_type
+    and self.job.mat_index == obj.iobj.mat_index
+    and self.job.job_type ~= df.job_type.PrepareMeal
+    then
+        self.job.mat_type = mat_type
+        self.job.mat_index = mat_index
+    end
+
+    obj.iobj.mat_type = mat_type
+    obj.iobj.mat_index = mat_index
+end
+
+function JobDetails:findUnambiguousItem(iobj)
+    local count = 0
+    local itype
+
+    for i = 0,df.item_type._last_item do
+        if dfhack.job.isSuitableItem(iobj, i, -1) then
+            count = count + 1
+            if count > 1 then return nil end
+            itype = i
+        end
+    end
+
+    return itype
+end
+
+function JobDetails:onChangeMat()
+    local idx, obj = self.list:getSelected()
+
+    if obj.iobj.item_type == -1 and obj.iobj.mat_type == -1 then
+        -- If the job allows only one specific item type, use it
+        local vitype = self:findUnambiguousItem(obj.iobj)
+
+        if vitype then
+            obj.iobj.item_type = vitype
+        else
+            dlg.showMessage(
+                'Bug Alert',
+                { 'Please set a specific item type first.\n\n',
+                  'Otherwise the material will be matched\n',
+                  'incorrectly due to a limitation in DF code.' },
+                COLOR_YELLOW
+            )
+            return
+        end
+    end
+
+    guimat.MaterialDialog{
+        prompt = 'Please select a new material for input '..idx,
+        none_caption = 'any material',
+        mat_filter = function(mat,parent,mat_type,mat_index)
+            return dfhack.job.isSuitableMaterial(obj.iobj, mat_type, mat_index, obj.iobj.item_type)
+        end,
+        on_select = self:callback('setMaterial', obj)
+    }:show()
+end
+
+function JobDetails:canChangeTrait()
+    local idx, obj = self.list:getSelected()
+    return obj ~= nil and not is_caste_mat(obj.iobj)
+end
+
+function JobDetails:onChangeTrait()
+    local idx, obj = self.list:getSelected()
+    guimat.ItemTraitsDialog{
+        job_item = obj.iobj,
+        prompt = 'Please select traits for input '..idx,
+        none_caption = 'no traits',
+    }:show()
+end
+
+local function ScrJobDetails()
+    return df.global.game.main_interface.job_details
+end
+
+local function show_job_details()
+    local scr = ScrJobDetails()
+    if not scr.open -- dfhack.gui.matchFocusString('dwarfmode/JobDetails')
+    then
+        qerror("This script needs to be run from a job details screen")
+    end
+
+    local job
+    if scr.context == df.job_details_context_type.BUILDING_TASK_LIST then
+        job = scr.jb
+    elseif scr.context == df.job_details_context_type.MANAGER_WORK_ORDER then
+        job = scr.wq
+    end
+    if job == nil then
+        qerror("Unhandled screen context: ".. df.job_details_context_type[scr.context])
+    end
+
+    local dlg = JobDetails{ job = job, context = scr.context }
+    dlg:show()
```

Comment:
naming conflict with global `dlg`

Reply:
but does the conflict matter? It's local to this scope, it doesn't break anything (I actually stored the value in a global variable in earlier iterations of the rework -- when pressing `ESC` wouldn't close the window)